File: CSharp/Legacy89DiskKit.FileSystem.Infrastructure/FileSystem/HuBasic/HuBasicWriteRules.cs
using Legacy89DiskKit.Domain.FileSystem.Model;
using Legacy89DiskKit.Infrastructure.FileSystem.HuBasic.Models;

namespace Legacy89DiskKit.Infrastructure.FileSystem.HuBasic;

public static class HuBasicWriteRules
{
    public static byte[] PrepareWritePayload(byte[] data, ExtendedFileAttributes attributes)
    {
        if (data == null) throw new ArgumentNullException(nameof(data));
        if (attributes == null) throw new ArgumentNullException(nameof(attributes));

        if (!attributes.IsAscii)
        {
            return data;
        }

        if (data.Length > 0 && data[^1] == 0x1A)
        {
            return data;
        }

        var newData = new byte[data.Length + 1];
        Array.Copy(data, newData, data.Length);
        newData[^1] = 0x1A;
        return newData;
    }

    public static int GetClustersNeeded(int dataLength, HuBasicConfiguration config)
    {
        if (config == null) throw new ArgumentNullException(nameof(config));

        int clustersNeeded = (dataLength + config.ClusterSize - 1) / config.ClusterSize;
        return clustersNeeded == 0 ? 1 : clustersNeeded;
    }

    public static int GetSectorsInLastCluster(int dataLength, HuBasicConfiguration config)
    {
        if (config == null) throw new ArgumentNullException(nameof(config));

        int sectorsPerCluster = config.ClusterSize / config.SectorSize;
        int sectorsInLastCluster = ((dataLength + config.SectorSize - 1) / config.SectorSize) % sectorsPerCluster;
        return sectorsInLastCluster == 0 ? sectorsPerCluster : sectorsInLastCluster;
    }

    public static int GetTerminalFlagForLength(int dataLength, HuBasicConfiguration config)
    {
        return 0x7F + GetSectorsInLastCluster(dataLength, config);
    }
}

File: Cpp/Legacy89DiskKit.Cpp/src/hu_basic_write_rules.cpp
#include "legacy89diskkit/cpp/hu_basic_write_rules.hpp"

namespace legacy89diskkit::cpp
{
std::vector<std::uint8_t> HuBasicWriteRules::PrepareWritePayload(const std::vector<std::uint8_t>& data, const HuBasicFileAttributes& attributes)
{
    if (!attributes.is_ascii)
    {
        return data;
    }

    if (!data.empty() && data.back() == 0x1a)
    {
        return data;
    }

    auto prepared = data;
    prepared.push_back(0x1a);
    return prepared;
}

int HuBasicWriteRules::GetClustersNeeded(int data_length, const HuBasicConfiguration& config)
{
    const auto clusters_needed = (data_length + config.cluster_size - 1) / config.cluster_size;
    return clusters_needed == 0 ? 1 : clusters_needed;
}

int HuBasicWriteRules::GetSectorsInLastCluster(int data_length, const HuBasicConfiguration& config)
{
    const auto sectors_per_cluster = config.cluster_size / config.sector_size;
    const auto sectors_in_last_cluster = ((data_length + config.sector_size - 1) / config.sector_size) % sectors_per_cluster;
    return sectors_in_last_cluster == 0 ? sectors_per_cluster : sectors_in_last_cluster;
}

int HuBasicWriteRules::GetTerminalFlagForLength(int data_length, const HuBasicConfiguration& config)
{
    return 0x7f + GetSectorsInLastCluster(data_length, config);
}
}
